src/bitprim/keoken/interpreter.cpp

```diff
@@ -0,0 +1,473 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <bitprim/keoken/interpreter.hpp>
+
+#include <bitcoin/bitcoin/chain/output.hpp>
+#include <bitcoin/bitcoin/utility/container_source.hpp>
+#include <bitcoin/bitcoin/utility/istream_reader.hpp>
+
+#include <bitprim/keoken/transaction_extractor.hpp>
+
+using libbitcoin::chain::transaction;
+using libbitcoin::chain::output;
+using libbitcoin::data_chunk;
+using libbitcoin::reader;
+using libbitcoin::wallet::payment_address;
+
+namespace bitprim {
+namespace keoken {
+
+using error::error_code_t;
+
+// explicit
+interpreter::interpreter(libbitcoin::blockchain::fast_chain& fast_chain, state& st)
+    : fast_chain_(fast_chain)
+    , state_(st)
+{}
+
+error_code_t interpreter::process(size_t block_height, transaction const& tx) {
+    using libbitcoin::istream_reader;
+    using libbitcoin::data_source;
+
+    auto data = first_keoken_output(tx);
+    if ( ! data.empty()) {
+        data_source ds(data);
+        istream_reader source(ds);
+
+        return version_dispatcher(block_height, tx, source);
+    }
+    return error::not_keoken_tx;
+}
+
+error_code_t interpreter::version_dispatcher(size_t block_height, transaction const& tx, reader& source) {
+
+    auto version = source.read_2_bytes_big_endian();
+    if ( ! source) return error::invalid_version_number;
+
+    switch (static_cast<version_t>(version)) {
+        case version_t::zero:
+            return version_0_type_dispatcher(block_height, tx, source);
+    }
+    return error::not_recognized_version_number;
+}
+
+error_code_t interpreter::version_0_type_dispatcher(size_t block_height, transaction const& tx, reader& source) {
+    auto type = source.read_2_bytes_big_endian();
+    if ( ! source) return error::invalid_type;
+
+    switch (static_cast<message_type_t>(type)) {
+        case message_type_t::create_asset:
+            return process_create_asset_version_0(block_height, tx, source);
+        case message_type_t::send_tokens:
+            return process_send_tokens_version_0(block_height, tx, source);
+    }
+    return error::not_recognized_type;
+}
+
+payment_address interpreter::get_first_input_addr(transaction const& tx) const {
+    auto const& owner_input = tx.inputs()[0];
+
+    output out_output;
+    size_t out_height;
+    uint32_t out_median_time_past;
+    bool out_coinbase;
+
+    if ( ! fast_chain_.get_output(out_output, out_height, out_median_time_past, out_coinbase, 
+                                  owner_input.previous_output(), libbitcoin::max_size_t, true)) {
+        return payment_address{};
+    }
+
+    return out_output.address();
+}
+
+error_code_t interpreter::process_create_asset_version_0(size_t block_height, transaction const& tx, reader& source) {
+    auto msg = message::create_asset::factory_from_data(source);
```

Comment:
missing msg construction validation

Reply:
Fixed in `bitprim-core` 
https://github.com/bitprim/bitprim-core/pull/37